### DISCUSSION TITLE
Algorithm details: The initPlacement step of gpl(src/gpl/src/initialPlace.cpp) miss coefficient "2" in "-weightY  * ((pin1->cy() - pin1->instance()->cy()) - (pin2->cy() - pin2->instance()->cy()))"

### DISCUSSION THREAD

**Discussion by TuringKi** [The-OpenROAD-Project/OpenROAD#4707] [General]
From [1], we have the object function about wire length:

$$ \Gamma =\frac{1}{2}\sum_{i=1}^{M}\sum_{j=i + 1}^{M} \omega_{i,j}(x_i^{p} - x_j^{p})^2$$

where $M$ is the number of instances, and,

$$  \omega_{i,j} = \frac{2}{(M - 1)} \frac{1}{|x_i^{p} - x_j^{p}|}, $$

for $x_i^{p}$ or $x_j^{p}$ are boarder pins, otherwise $\omega_{i,j} =0$.

The definition of  $\omega_{i,j}$ makes $\Gamma$ equal exactly the HPWL [1].

Let $\pi_i$ be the center point of instance $i$, and $o_i=\pi_i - x_i^p$, then:

$$ \Gamma =\frac{1}{2}\sum_{i=1}^{M}\sum_{j=i + 1}^{M} \omega_{i,j}( \pi_i - o_i  - \pi_j + o_j )^2 =\frac{1}{2}\sum_{i=1}^{M}\sum_{j=i + 1}^{M} \omega_{i,j} [ (\pi_i - \pi_j)^2 - 2\pi_i(o_i - o_j) - 2\pi_j(o_j - o_i) + (o_i - o_j)^2 ]$$

where $(o_i - o_j)^2$ is constant, and can be omitted in gradient.

Suppose all the instances are movable cells, let $\text{x} = [\pi_1,\pi_2,...,\pi_M]^T$ convert it to matrix format:

$$  \Gamma =\text{x}^T\text{C}\text{x} + \text{b}^T\text{x} + constant $$

and, 

$$ \text{b}_i = - 2\omega(o_i - o_j),  \text{b}_j = - 2 \omega (o_j - o_i)  $$

However, in (https://github.com/The-OpenROAD-Project/OpenROAD/blob/4afdd47eb2f8103c2832951f9c8f2b4a11d37bc1/src/gpl/src/initialPlace.cpp#L378), its:

```c++
 fixedInstForceVecY_(inst1)
                += -weightY
                   * ((pin1->cy() - pin1->instance()->cy())
                      - (pin2->cy() - pin2->instance()->cy()));
            fixedInstForceVecY_(inst2)
                += -weightY
                   * ((pin2->cy() - pin2->instance()->cy())
                      - (pin1->cy() - pin1->instance()->cy()));
```

Reference:
1. P. Spindler, U. Schlichtmann, and F. M. Johannes. 2008. Kraftwerk2 - A Fast Force-Directed Quadratic Placement Approach Using an Accurate Net Model. IEEE TCAD 27, 8 (2008), 1398–1411

**Comment by maliberty** [comment]
Thanks for the detailed analysis.  Given the arbitrary weight in 
```
    float netWeight = ipVars_.netWeightScale / (net->pins().size() - 1);
```
with 
```
  initialPlaceNetWeightScale_ = 800;
```
I'm not sure the factor of two makes a difference.    We could add the two here and cut the other weight to 400.  The more interesting question is where 800 came from and is it a good value.  The commit is from @mgwoo in https://github.com/The-OpenROAD-Project/RePlAce/commit/07d7bdf382423b204ccd2cfc66b49293cbea64f6 but there is little explanation.

**Reply by mgwoo** [reply]
Original standalone RePlAce uses 1.0 as ipVars_.netWeightScale value. (See https://github.com/The-OpenROAD-Project/RePlAce/blob/7c07cffdf419c4dbd315c73c235ec5d670fb0240/src/initPlacement.cpp#L495) 
I remember the 1.0 worked well with academic bookshelf designs, but in lef/def scale, the initial place engine diverged.
As long as the QoR is okay and the BiCGSTAB solver is stable, this parameter can be tuned.
